### PR TITLE
Add normative requirement to prevent denial-of-service attacks

### LINKED ISSUE #120: What a reasonable way of defending against poison graphs?
This is more a question to those who have a practical experience of the algorithm implementation out there, ie, @davidlehn @dlongley and maybe @gkellogg.

As a protection against poison graphs what I did in my implementation is as follows:

1. I have set a maximum limit to the recursion in [5.4.5.1.](https://www.w3.org/TR/rdf-canon/#hndq.5.4.5.1). I.e., if the "level" of the recursion has reached a limit, I stop the algorithm. At this moment, the limit is set to 50, but I must admit this more a gut feeling rather than anything else.
2. The user has the possibility, via a separate call in the interface, to set a _lower_ maximum limit (but cannot exceed that limit).

Is this a reasonable approach? If yes, is 50 indeed a reasonable limit?

An alternative would be to let the system-wide limit be set at positive infinity (more exactly, the maximum integer representable in JavaScript), and the user can set a limit by herself.

Any advice/experience is welcome...

### PR DESCRIPTION
**Pull request by gkellogg**
... via poison datasets.

Fixes #120.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-canon/pull/138.html" title="Last updated on Jul 14, 2023, 3:19 PM UTC (9098bfe)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/rdf-canon/138/5c8e5e6...9098bfe.html" title="Last updated on Jul 14, 2023, 3:19 PM UTC (9098bfe)">Diff</a>

### CODE REVIEW

**Comment by gkellogg** [comment]
I think this satisfies the requirement without being overly prescriptive.

File: spec/index.html [hunk 1]
```diff
@@ -980,6 +980,18 @@ <h3>Examples</h3>
     <section id="canon-algo-algo">
       <h3>Algorithm</h3>
 
+      <p>The following algorithm will run with a minimal number of iterations in each step
+        for well formed <a>input datasets</a>.
```
Comment (iherman) [inline spec/index.html:984]: What does "well formed" means? Poison sets are also well formed, i.e., they are bona fide RDF graph datasets...

"average"? "general"? "practical?"

File: spec/index.html [hunk 2]
```diff
@@ -980,6 +980,18 @@ <h3>Examples</h3>
...
+        In some extreme cases, the algorihm can behave poorly, particularly in <a href="#ca.5">Step 5</a>.
+        Implementations MUST prevent against potenial denial-of-service attacks.
+        See <a href="#dataset-poisoning"></a> for futher information.</p>
+
+      <p class="note">Implementations can consider placing limits on the number of
+        calls to <a href="#hash-nd-quads"></a> based on the number
+        of blank nodes in the <a>hash to blank nodes map</a>.
+        For most well-formed datasets, no more than a couple
+        of iterations on <a href="#hash-nd-quads"></a> per blank node would be unusual.</p>
```
Comment (iherman) [inline spec/index.html:993]:
```suggestion
        For most well-formed datasets, more than a couple
        of iterations on <a href="#hash-nd-quads"></a> per blank node would be unusual.</p>
```

Note also my comment above on the term "well-formed".

(Minor: even if the term is retained, it should be consistent on whether it is "well-formed" or "well formed".)

**Review by iherman — CHANGES_REQUESTED** [review]

File: spec/index.html (hunk 1 above)
Comment (gkellogg) [reply spec/index.html:984]: Maybe. “Well behaved” or “typical”.

Comment (iherman) [reply spec/index.html:984]: I'd vote for "typical"

**Review by TallTed — CHANGES_REQUESTED** [review]
I would change both instances of `well[- ]formed` to `typical`.

**Review by dlongley — APPROVED** [review]
+1 with the suggestions offered by others.

File: spec/index.html (hunk 1 above)
Comment (gkellogg) [reply spec/index.html:984]:
```suggestion
        for typical <a>input datasets</a>.
```

**Review by iherman — APPROVED** [review]

**Pull request merged**